### PR TITLE
Structure API Update

### PR DESCRIPTION
**Pull request by UnderscoreTud**
Convert code using SelfRegisteringSkriptEvent to use Structures and start moving classes into the new package (`org.skriptlang.skript`)

Fixes https://github.com/SkriptLang/skript-reflect/issues/75
Fixes https://github.com/SkriptLang/skript-reflect/issues/72
Fixes https://github.com/SkriptLang/skript-reflect/issues/69

### CODE REVIEW

**Review by Moderocky — APPROVED** [review]

**Review by sovdeeth — APPROVED** [review]
my eyes glazed over halfway through the code but all my testing couldn't produce any bugs
so seems good for dev build testing

File: src/main/java/com/btk5h/skriptmirror/JavaType.java [hunk 1]
```diff
@@ -2,7 +2,8 @@
 
 import java.util.Objects;
 
-public final class JavaType {
+public class JavaType {
```
Comment (APickledWalrus) [inline src/main/java/com/btk5h/skriptmirror/JavaType.java]: Any reason for removing the final modifier?

File: src/main/java/com/btk5h/skriptmirror/skript/custom/SyntaxParseEvent.java [hunk 2]
```diff
@@ -54,15 +56,14 @@ public HandlerList getHandlers() {
...
-                                                                         List<T> whichInfo, Map<T, Trigger> parserHandlers) {
+  public static <T extends CustomSyntaxStructure.SyntaxData> void register(SectionNode parseNode,
+                                                                           List<T> whichInfo, Map<T, Trigger> parserHandlers) {
     ParserInstance.get().setCurrentEvent("custom syntax parser", SyntaxParseEvent.class);
     List<TriggerItem> items = SkriptUtil.getItemsFromNode(parseNode);
 
     whichInfo.forEach(which ->
         parserHandlers.put(which,
-            new Trigger(ParserInstance.get().getCurrentScript(), "parse " + which.getPattern(), section, items)));
+            new Trigger(ParserInstance.get().getCurrentScript(), "parse " + which.getPattern(), new SimpleEvent(), items)));
```
Comment (APickledWalrus) [inline src/main/java/com/btk5h/skriptmirror/skript/custom/SyntaxParseEvent.java:66]: Is this "safe"? I see it used to take in a CustomSyntaxSection. Can it take in a CustomSyntaxStructure now?

File: src/main/java/org/skriptlang/reflect/java/elements/structures/StructImport.java [hunk 3]
```diff
@@ -23,87 +18,75 @@
...
+import org.skriptlang.skript.lang.script.Script;
+import org.skriptlang.skript.lang.structure.Structure;
 
-import java.io.File;
-import java.util.HashMap;
-import java.util.Map;
-import java.util.Optional;
-import java.util.Spliterator;
-import java.util.Spliterators;
+import java.util.*;
```
Comment (APickledWalrus) [inline src/main/java/org/skriptlang/reflect/java/elements/structures/StructImport.java]: ;)

File: src/main/java/org/skriptlang/reflect/java/elements/structures/StructImport.java [hunk 4]
```diff
@@ -23,87 +18,75 @@
...
-    public boolean init(Literal<?>[] args, int matchedPattern, ParseResult parseResult) {
-      File currentScript = SkriptUtil.getCurrentScriptFile();
-      SectionNode node = (SectionNode) SkriptLogger.getNode();
-
-      if (node.getKey().toLowerCase().startsWith("on ")) {
-        return false;
-      }
+  @Override
+  public boolean init(Literal<?>[] args, int matchedPattern, ParseResult parseResult, EntryContainer entryContainer) {
+    this.script = SkriptUtil.getCurrentScript();
```
Comment (APickledWalrus) [inline src/main/java/org/skriptlang/reflect/java/elements/structures/StructImport.java]: it should be safe (and more efficient) to use `getParser().getCurrentScript()` here

File: src/main/java/org/skriptlang/reflect/syntax/condition/elements/StructCustomCondition.java [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+
+public class StructCustomCondition extends CustomSyntaxStructure<ConditionSyntaxInfo> {
+
+  public static boolean customConditionsUsed = false;
+
+  static {
+    String[] syntax = {
+        "[(1¦local)] condition <.+>",
+        "[(1¦local)] condition",
+        "[(1¦local)] %*classinfos% property condition <.+>"
```
Comment (APickledWalrus) [inline src/main/java/org/skriptlang/reflect/syntax/condition/elements/StructCustomCondition.java]:
```suggestion
        "[1:local] condition <.+>",
        "[1:local] condition",
        "[1:local] %*classinfos% property condition <.+>"
```
Ideally we could use parse tags here too :)

File: src/main/java/org/skriptlang/reflect/syntax/condition/elements/StructCustomCondition.java [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+  @Override
+  public boolean init(Literal<?>[] args, int matchedPattern, SkriptParser.ParseResult parseResult,
+                      EntryContainer entryContainer) {
+    customConditionsUsed = true;
+
+    SectionNode patterns = entryContainer.getOptional("patterns", SectionNode.class, false);
+    Script script = (parseResult.mark & 1) == 1 ? SkriptUtil.getCurrentScript() : null;
+
+    switch (matchedPattern) {
+      case 0:
```
Comment (APickledWalrus) [inline src/main/java/org/skriptlang/reflect/syntax/condition/elements/StructCustomCondition.java]:
```suggestion
      case 0: // condition with inline patterns
```
comments explaining each case would be good for readability

File: src/main/java/org/skriptlang/reflect/syntax/condition/elements/StructCustomCondition.java [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+        register = () -> {
+          register(ConditionSyntaxInfo.create(script, "%" + type + "% (is|are) " + property, 1, false, true));
+          register(
+              ConditionSyntaxInfo.create(script, "%" + type + "% (isn't|is not|aren't|are not) " + property, 1, true, true));
+        };
+        break;
+    }
+
+    if (matchedPattern != 1 && patterns != null) {
+      Skript.error("Custom conditions with inline patterns may not have a patterns section.");
```
Comment (APickledWalrus) [inline src/main/java/org/skriptlang/reflect/syntax/condition/elements/StructCustomCondition.java]:
```suggestion
      Skript.error("A custom condition with an inline pattern cannot have a 'patterns' entry too.");
```

File: src/main/java/org/skriptlang/reflect/syntax/condition/elements/StructCustomCondition.java [hunk 8]
```diff
@@ -0,0 +1,186 @@
...
+  }
+
+  @SuppressWarnings("unchecked")
+  @Override
+  public boolean init(Literal<?>[] args, int matchedPattern, SkriptParser.ParseResult parseResult,
+                      EntryContainer entryContainer) {
+    customConditionsUsed = true;
+
+    SectionNode patterns = entryContainer.getOptional("patterns", SectionNode.class, false);
+    Script script = (parseResult.mark & 1) == 1 ? SkriptUtil.getCurrentScript() : null;
```
Comment (APickledWalrus) [inline src/main/java/org/skriptlang/reflect/syntax/condition/elements/StructCustomCondition.java]:
```suggestion
    Script script = parseResult.mark == 1 ? SkriptUtil.getCurrentScript() : null;
```
I do not believe we need to perform this operation? There is only 1 parse mark

File: src/main/java/org/skriptlang/reflect/syntax/condition/elements/StructCustomCondition.java [hunk 9]
```diff
@@ -0,0 +1,186 @@
...
+
+    switch (matchedPattern) {
+      case 0:
+        String pattern = parseResult.regexes.get(0).group();
+        register = () -> register(ConditionSyntaxInfo.create(script, pattern, 1, false, false));
+        break;
+      case 1:
+        if (patterns == null) {
+          Skript.error("Custom conditions without inline patterns must have a patterns section.");
```
Comment (APickledWalrus) [inline src/main/java/org/skriptlang/reflect/syntax/condition/elements/StructCustomCondition.java]:
```suggestion
          Skript.error("A custom condition without an inline pattern must have a 'patterns' entry.");
```

File: src/main/java/org/skriptlang/reflect/syntax/condition/elements/StructCustomCondition.java [hunk 10]
```diff
@@ -0,0 +1,186 @@
...
+  @Override
+  public boolean preLoad() {
+    super.preLoad();
+    register.run();
+    EntryContainer entryContainer = getEntryContainer();
+    parseNode = entryContainer.getOptional("parse", SectionNode.class, false);
+    SectionNode safeParseNode = entryContainer.getOptional("safe parse", SectionNode.class, false);
+    if (parseNode != null) {
+      if (safeParseNode != null) {
+        Skript.error("You can't have two parse sections");
```
Comment (APickledWalrus) [inline src/main/java/org/skriptlang/reflect/syntax/condition/elements/StructCustomCondition.java]:
```suggestion
        Skript.error("A custom syntax element cannot contain both 'parse' and 'safe parse' entries");
```
It seems that this part of the code is repeated in each Structure. Is there any way we can abstract this out?

File: src/main/java/org/skriptlang/reflect/syntax/condition/elements/StructCustomCondition.java [hunk 11]
```diff
@@ -0,0 +1,186 @@
...
+        whichInfo.forEach(which -> parseSectionLoaded.put(which, true));
+    }
+
+    SectionNode usableInNode = entryContainer.getOptional("usable in", SectionNode.class, false);
+    if (usableInNode != null && !handleUsableSection(usableInNode, usableSuppliers))
+      return false;
+
+    checkNode = entryContainer.getOptional("check", SectionNode.class, false);
+    if (checkNode == null)
+      Skript.warning("Custom conditions are useless without a check section");
```
Comment (APickledWalrus) [inline src/main/java/org/skriptlang/reflect/syntax/condition/elements/StructCustomCondition.java]: Why even allow this? IMO we should just make the entry required

File: src/main/java/org/skriptlang/reflect/syntax/condition/elements/StructCustomCondition.java [hunk 12]
```diff
@@ -0,0 +1,186 @@
...
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+
+public class StructCustomCondition extends CustomSyntaxStructure<ConditionSyntaxInfo> {
+
+  public static boolean customConditionsUsed = false;
```
Comment (APickledWalrus) [inline src/main/java/org/skriptlang/reflect/syntax/condition/elements/StructCustomCondition.java:34]: Where is this used? It probably needs to be updated to support unloading, right?

File: src/main/java/org/skriptlang/reflect/syntax/event/elements/StructCustomEvent.java [hunk 13]
```diff
@@ -0,0 +1,186 @@
...
+
+  public static boolean customEventsUsed = false;
+
+  static {
+    String[] syntax = {
+      "[(1¦local)] [custom] event %string%"
+    };
+    Skript.registerStructure(StructCustomEvent.class, customSyntaxValidator()
+        .addEntry("pattern", null, true)
+        .addEntry("event-values", null, true)
```
Comment (APickledWalrus) [inline src/main/java/org/skriptlang/reflect/syntax/event/elements/StructCustomEvent.java]: What do you think about supporting `event values`? Would require some additional work to keep backwards compatibility but this is a weird design choice.

File: src/main/java/org/skriptlang/reflect/syntax/condition/elements/StructCustomCondition.java [hunk 14]
```diff
@@ -0,0 +1,186 @@
...
+        break;
+      case 1:
+        if (patterns == null) {
+          Skript.error("Custom conditions without inline patterns must have a patterns section.");
+          return false;
+        }
+
+        register = () -> {
+          int i = 1;
+          for (Node subNode : patterns) {
```
Comment (APickledWalrus) [inline src/main/java/org/skriptlang/reflect/syntax/condition/elements/StructCustomCondition.java]: This does not check for there being no subnodes, while other custom syntax implementations do. Maybe you could add a utility method to CustomSyntaxStructure?

File: src/main/java/org/skriptlang/reflect/syntax/expression/ChangerEntryData.java [hunk 15]
```diff
@@ -0,0 +1,58 @@
...
+import java.util.Arrays;
+
+public class ChangerEntryData extends EntryData<NonNullPair<SectionNode, Class<?>[]>> {
+
+  public ChangerEntryData(String key, boolean optional) {
+    super(key, null, optional);
+  }
+
+  @Override
+  public @Nullable NonNullPair<SectionNode, Class<?>[]> getValue(Node node) {
```
Comment (APickledWalrus) [inline src/main/java/org/skriptlang/reflect/syntax/expression/ChangerEntryData.java]:
```suggestion
  @Nullable
  public NonNullPair<SectionNode, Class<?>[]> getValue(Node node) {
```

**Review by APickledWalrus — CHANGES_REQUESTED** [review]
I have not yet taken too deep of look into the other custom syntax structures, but they have the same "issues" as StructCustomCondition

excellent work so far!

File: src/main/java/org/skriptlang/reflect/syntax/condition/elements/StructCustomCondition.java (hunk 12 above)
Comment (UnderscoreTud) [reply src/main/java/org/skriptlang/reflect/syntax/condition/elements/StructCustomCondition.java:34]: Only for bStats.

File: src/main/java/org/skriptlang/reflect/syntax/event/elements/StructCustomEvent.java (hunk 13 above)
Comment (UnderscoreTud) [reply src/main/java/org/skriptlang/reflect/syntax/event/elements/StructCustomEvent.java]: I will look into it

File: src/main/java/org/skriptlang/reflect/syntax/condition/elements/StructCustomCondition.java (hunk 5 above)
Comment (abandonedaccount6235) [reply src/main/java/org/skriptlang/reflect/syntax/condition/elements/StructCustomCondition.java]: this is definitely a good idea, but maybe we can open an issue for it and leave it as is for now in the interest of a sooner release

File: src/main/java/org/skriptlang/reflect/syntax/event/elements/StructCustomEvent.java (hunk 13 above)
Comment (abandonedaccount6235) [reply src/main/java/org/skriptlang/reflect/syntax/event/elements/StructCustomEvent.java]: maybe this is another thing that could be an issue for the future

File: src/main/java/org/skriptlang/reflect/syntax/condition/elements/StructCustomCondition.java (hunk 5 above)
Comment (APickledWalrus) [reply src/main/java/org/skriptlang/reflect/syntax/condition/elements/StructCustomCondition.java]: parse tags can wait for sure, though I think applying the suggested change would be good as it requires no other code changes

File: src/main/java/org/skriptlang/reflect/syntax/CustomSyntaxStructure.java [hunk 16]
```diff
@@ -285,20 +182,41 @@ private void update() {
...
+    SectionNode safeParseNode = getEntryContainer().getOptional("safe parse", SectionNode.class, false);
+    if (parseNode != null && safeParseNode != null) {
+      Skript.error("A custom syntax element cannot contain both 'parse' and 'safe parse' entries");
+      return null;
+    }
+    if (safeParseNode != null) {
+      Skript.warning("The 'safe parse' entry is deprecated and will act as a regular 'parse' entry."
+          + " Please use the 'parse' entry instead");
+      return new SectionNode[] {safeParseNode};
+    }
```
Comment (APickledWalrus) [inline src/main/java/org/skriptlang/reflect/syntax/CustomSyntaxStructure.java]:
```suggestion
    if (safeParseNode != null) {
      if (parseNode != null {
        Skript.error("A custom syntax element cannot contain both 'parse' and 'safe parse' entries");
        return null;
      } else {
        Skript.warning("The 'safe parse' entry is deprecated and will act as a regular 'parse' entry."
          + " Please use the 'parse' entry instead");
        return new SectionNode[] {safeParseNode};
      }
    }
```
maybe this is better

File: src/main/java/org/skriptlang/reflect/syntax/PatternsEntryData.java [hunk 17]
```diff
@@ -0,0 +1,41 @@
...
+import ch.njol.skript.ScriptLoader;
+import ch.njol.skript.config.Node;
+import ch.njol.skript.config.SectionNode;
+import org.eclipse.jdt.annotation.Nullable;
+import org.skriptlang.skript.lang.entry.EntryData;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class PatternsEntryData extends EntryData<List<String>> {
```
Comment (APickledWalrus) [inline src/main/java/org/skriptlang/reflect/syntax/PatternsEntryData.java]: maybe package private?

File: src/main/java/org/skriptlang/reflect/syntax/condition/elements/StructCustomCondition.java [hunk 18]
```diff
@@ -0,0 +1,166 @@
...
+
+  static {
+    String[] syntax = {
+        "[:local] condition <.+>",
+        "[:local] condition",
+        "[:local] %*classinfos% property condition <.+>"
+    };
+    Skript.registerStructure(StructCustomCondition.class, customSyntaxValidator()
+        .addSection("check", false)
+        .build(), syntax);
```
Comment (APickledWalrus) [inline src/main/java/org/skriptlang/reflect/syntax/condition/elements/StructCustomCondition.java]:
```suggestion
    Skript.registerStructure(StructCustomCondition.class, customSyntaxValidator()
        .addSection("check", false)
        .build(),
        syntax
    );
```

File: src/main/java/org/skriptlang/reflect/syntax/condition/elements/StructCustomCondition.java [hunk 19]
```diff
@@ -0,0 +1,166 @@
...
+      return false;
+    }
+
+    switch (matchedPattern) {
+      case 0: // condition with an inline pattern
+        String pattern = parseResult.regexes.get(0).group();
+        register(ConditionSyntaxInfo.create(script, pattern, 1, false, false));
+        break;
+      case 1: // condition with a 'patterns' entry
+        if (patterns == null) {
```
Comment (APickledWalrus) [inline src/main/java/org/skriptlang/reflect/syntax/condition/elements/StructCustomCondition.java:96]: Isn't this what the checkHasPatterns() is for?

File: src/main/java/org/skriptlang/reflect/syntax/expression/elements/StructCustomExpression.java [hunk 20]
```diff
@@ -0,0 +1,269 @@
...
+  static {
+    String[] syntax = {
+      "[:local] [plural:(plural|non[-| ]single)] expression <.+>",
+      "[:local] [plural:(plural|non[-| ]single)] expression",
+      "[:local] [plural:(plural|non[-| ]single)] %*classinfos% property <.+>"
+    };
+    EntryValidator.EntryValidatorBuilder builder = customSyntaxValidator()
+        .addEntryData(new KeyValueEntryData<Class<?>>("return type", null, true) {
+          @Override
+          protected @Nullable Class<?> getValue(String value) {
```
Comment (APickledWalrus) [inline src/main/java/org/skriptlang/reflect/syntax/expression/elements/StructCustomExpression.java]:
```suggestion
          @Nullable
          protected Class<?> getValue(String value) {
```

**Review by APickledWalrus — CHANGES_REQUESTED** [review]
looking pretty good

File: src/main/java/org/skriptlang/reflect/syntax/condition/elements/StructCustomCondition.java (hunk 19 above)
Comment (UnderscoreTud) [reply src/main/java/org/skriptlang/reflect/syntax/condition/elements/StructCustomCondition.java:96]: This is for a more specific error

Comment (APickledWalrus) [reply src/main/java/org/skriptlang/reflect/syntax/condition/elements/StructCustomCondition.java:96]: Is there a purpose for the checkHasPatterns method then?

Comment (UnderscoreTud) [reply src/main/java/org/skriptlang/reflect/syntax/condition/elements/StructCustomCondition.java:96]: For a more general use case. For example if the `patterns` entry is present but empty.

**Comment by envizar** [comment]
Is there already a stable version with these changes?
New version of Hippo depends on skript-reflect with these changes

File: src/main/java/org/skriptlang/reflect/java/elements/structures/StructImport.java [hunk 21]
```diff
@@ -45,8 +45,7 @@ public class StructImport extends Structure {
 
     // TODO try replacing ImportHandler with JavaType's literal parsing
     Skript.registerExpression(ImportHandler.class, JavaType.class, ExpressionType.SIMPLE);
-    thisInfo = StreamSupport.stream(
-            Spliterators.spliteratorUnknownSize(Skript.getExpressions(), Spliterator.ORDERED), false)
+    thisInfo = StreamSupport.stream(Spliterators.spliteratorUnknownSize(Skript.getExpressions(), Spliterator.ORDERED), false)
         .filter(i -> i.c == ImportHandler.class)
```
Comment (Moderocky) [inline src/main/java/org/skriptlang/reflect/java/elements/structures/StructImport.java]: Do you think we can replace some of these single-letter variable names as we go through? It's impossible to tell what this is without opening it in an IDE.

**Review by Moderocky — APPROVED** [review]

File: src/main/java/org/skriptlang/reflect/java/elements/structures/StructImport.java [hunk 22]
```diff
@@ -34,76 +31,66 @@
...
-
-    if (info.isPresent()) {
-      thisInfo = info.get();
-    } else {
-      Skript.warning("Could not find custom import class. Custom imports will not work.");
-    }
+    thisInfo = StreamSupport.stream(Spliterators.spliteratorUnknownSize(Skript.getExpressions(), Spliterator.ORDERED), false)
+        .filter(expressionInfo -> ImportHandler.class.equals(expressionInfo.getElementClass()))
+        .findFirst().orElseGet(() -> {
+          Skript.warning("Could not find custom import class. Custom imports will not work.");
```
Comment (APickledWalrus) [inline src/main/java/org/skriptlang/reflect/java/elements/structures/StructImport.java]: Would this ever be the case? Why do we not do this check for the other custom element classes?

File: src/main/java/org/skriptlang/reflect/syntax/condition/elements/StructCustomCondition.java [hunk 23]
```diff
@@ -0,0 +1,168 @@
...
+              if (!isPlural) {
+                return Utils.toEnglishPlural(codeName);
+              }
+
+              return codeName;
+            })
+            .collect(Collectors.joining("/"));
+        register(ConditionSyntaxInfo.create(script, "%" + type + "% (is|are) " + property, 1, false, true));
+        register(
+            ConditionSyntaxInfo.create(script, "%" + type + "% (isn't|is not|aren't|are not) " + property, 1, true, true));
```
Comment (APickledWalrus) [inline src/main/java/org/skriptlang/reflect/syntax/condition/elements/StructCustomCondition.java]:
```suggestion
        register(ConditionSyntaxInfo.create(script, "%" + type + "% (isn't|is not|aren't|are not) " + property, 1, true, true));
```
or
```suggestion
        register(
            ConditionSyntaxInfo.create(script, "%" + type + "% (isn't|is not|aren't|are not) " + property, 1, true, true)
        );
```

File: src/main/java/org/skriptlang/reflect/syntax/condition/elements/StructCustomCondition.java [hunk 24]
```diff
@@ -0,0 +1,168 @@
...
+
+      whichInfo.forEach(which -> parseSectionLoaded.put(which, true));
+    }
+
+    SectionNode checkNode = getEntryContainer().get("check", SectionNode.class, false);
+    SkriptLogger.setNode(checkNode);
+    getParser().setCurrentEvent("custom condition check", ConditionCheckEvent.class);
+    List<TriggerItem> items = SkriptUtil.getItemsFromNode(checkNode);
+    whichInfo.forEach(which -> conditionHandlers.put(which,
+        new Trigger(getParser().getCurrentScript(), "condition " + which, new SimpleEvent(), items)));
```
Comment (APickledWalrus) [inline src/main/java/org/skriptlang/reflect/syntax/condition/elements/StructCustomCondition.java]:
```suggestion
    whichInfo.forEach(which -> conditionHandlers.put(which,
        new Trigger(getParser().getCurrentScript(), "condition " + which, new SimpleEvent(), items))
    );
```

File: src/main/java/org/skriptlang/reflect/syntax/effect/elements/StructCustomEffect.java [hunk 25]
```diff
@@ -0,0 +1,144 @@
...
+      whichInfo.forEach(which -> parseSectionLoaded.put(which, true));
+    }
+
+    SectionNode triggerNode = getEntryContainer().get("trigger", SectionNode.class, false);
+    SkriptLogger.setNode(triggerNode);
+    getParser().setCurrentEvent("custom effect trigger", EffectTriggerEvent.class);
+    List<TriggerItem> items = SkriptUtil.getItemsFromNode(triggerNode);
+    whichInfo.forEach(which ->
+        effectHandlers.put(which,
+            new Trigger(getParser().getCurrentScript(), "effect " + which, new SimpleEvent(), items)));
```
Comment (APickledWalrus) [inline src/main/java/org/skriptlang/reflect/syntax/effect/elements/StructCustomEffect.java]:
```suggestion
        effectHandlers.put(which,
            new Trigger(getParser().getCurrentScript(), "effect " + which, new SimpleEvent(), items))
        );
```

File: src/main/java/org/skriptlang/reflect/syntax/event/elements/StructCustomEvent.java [hunk 26]
```diff
@@ -0,0 +1,165 @@
...
+import java.util.*;
+
+public class StructCustomEvent extends CustomSyntaxStructure<EventSyntaxInfo> {
+
+  public static boolean customEventsUsed = false;
+
+  static {
+    String[] syntax = {
+      "[:local] [custom] event %string%"
+    };
```
Comment (APickledWalrus) [inline src/main/java/org/skriptlang/reflect/syntax/event/elements/StructCustomEvent.java]: is there a reason for this?

File: src/main/java/org/skriptlang/reflect/syntax/event/elements/StructCustomEvent.java [hunk 27]
```diff
@@ -0,0 +1,165 @@
...
+    if (checkNode != null) {
+      SkriptLogger.setNode(checkNode);
+
+      getParser().setCurrentEvent("custom event trigger", EventTriggerEvent.class);
+      CustomEvent.setLastWhich(whichInfo.get(0));
+      List<TriggerItem> items = SkriptUtil.getItemsFromNode(checkNode);
+      CustomEvent.setLastWhich(null);
+      whichInfo.forEach(which ->
+          eventHandlers.put(which,
+              new Trigger(getParser().getCurrentScript(), "event " + which, new SimpleEvent(), items)));
```
Comment (APickledWalrus) [inline src/main/java/org/skriptlang/reflect/syntax/event/elements/StructCustomEvent.java]:
```suggestion
          eventHandlers.put(which,
              new Trigger(getParser().getCurrentScript(), "event " + which, new SimpleEvent(), items))
          );
```

File: src/main/java/org/skriptlang/reflect/syntax/expression/elements/StructCustomExpression.java [hunk 28]
```diff
@@ -0,0 +1,270 @@
...
+      getParser().setCurrentEvent("custom expression changer", ExpressionChangeEvent.class);
+      List<TriggerItem> items = SkriptUtil.getItemsFromNode(node);
+      whichInfo.forEach(which -> {
+        Map<Changer.ChangeMode, Trigger> changerMap =
+            changerHandlers.computeIfAbsent(which, k -> new HashMap<>());
+
+        String name = changeMode.toString().toLowerCase(Locale.ENGLISH).replace('_', ' ');
+        changerMap.put(changeMode,
+            new Trigger(getParser().getCurrentScript(),
+                String.format("%s %s", name, which.getPattern()), new SimpleEvent(), items));
```
Comment (APickledWalrus) [inline src/main/java/org/skriptlang/reflect/syntax/expression/elements/StructCustomExpression.java]:
```suggestion
        changerMap.put(changeMode,
            new Trigger(getParser().getCurrentScript(),
                String.format("%s %s", name, which.getPattern()), new SimpleEvent(), items
            )
        );
```

File: src/main/java/org/skriptlang/reflect/syntax/expression/elements/StructCustomExpression.java [hunk 29]
```diff
@@ -0,0 +1,270 @@
...
+    }
+
+    SectionNode getNode = getEntryContainer().get("get", SectionNode.class, false);
+    SkriptLogger.setNode(getNode);
+    {
+      getParser().setCurrentEvent("custom expression getter", ExpressionGetEvent.class);
+      List<TriggerItem> items = SkriptUtil.getItemsFromNode(getNode);
+      whichInfo.forEach(which ->
+          expressionHandlers.put(which,
+              new Trigger(getParser().getCurrentScript(), "get " + which.getPattern(), new SimpleEvent(), items)));
```
Comment (APickledWalrus) [inline src/main/java/org/skriptlang/reflect/syntax/expression/elements/StructCustomExpression.java]:
```suggestion
          expressionHandlers.put(which,
              new Trigger(getParser().getCurrentScript(), "get " + which.getPattern(), new SimpleEvent(), items))
          );
```

File: src/main/java/org/skriptlang/reflect/syntax/expression/elements/StructCustomExpression.java [hunk 30]
```diff
@@ -0,0 +1,270 @@
...
+          protected Class<?> getValue(String value) {
+            Class<?> returnType = Classes.getClassFromUserInput(value);
+            if (returnType == null)
+              Skript.error("The given return type doesn't exist");
+            return returnType;
+          }
+        })
+        .addEntry("loop of", null, true)
+        .addSection("get", false);
+    Arrays.stream(Changer.ChangeMode.values())
```
Comment (APickledWalrus) [inline src/main/java/org/skriptlang/reflect/syntax/expression/elements/StructCustomExpression.java]: Is there a reason for not directly importing ChangeMode?

**Review by APickledWalrus — APPROVED** [review]
Some final formatting suggestions. Excellent work! Next stop: merging :)

File: src/main/java/org/skriptlang/reflect/java/elements/structures/StructImport.java (hunk 22 above)
Comment (UnderscoreTud) [reply src/main/java/org/skriptlang/reflect/java/elements/structures/StructImport.java]: I'm not sure. Should I remove it from here or add it to the custom element classes?

Comment (APickledWalrus) [reply src/main/java/org/skriptlang/reflect/java/elements/structures/StructImport.java]: IMO either remove it or add a comment like "TODO figure out why this is here" 🤣

**Comment by rudde0** [comment]
idk is it correct place or not but I want to mention that skript-reflects detects `new ArrayList()` as skript function and it throws `The function 'ArrayList' does not exist.` error.
```v
import:
    java.util.ArrayList
    java.util.Collections
    java.util.Map$Entry
    ch.njol.skript.variables.Variables

# Example
#set {_topkillers::*} to the top (size of {Kills::*}) values of {Kills::*} as "&e@place. &a@index &7- &b@value"
expression replace values %strings% with %strings% in %string%:
    get:
        set {_string} to expression-3
        loop expressions-1:
            add 1 to {_index}
            replace all "%loop-value%" with "%{_index}th element of expressions-2%" in {_string}
        return {_string}
 
plural expression [the ](1¦(highest|top)|2¦(lowest|last)) %integer% values of %objects% [formatted ]as %string%:
    get:
        set {_list} to new ArrayList(Variables.getVariable(raw expressions-2.getName().toString(event), event, raw expressions-2.isLocal()).entrySet())
        ({_list}).sort(Entry.comparingByValue())
        if parse mark = 1:
            Collections.reverse({_list})
        loop ...{_list}:
            add 1 to {_index}
            set {_sorted::%{_index}%} to replace values "@place", "@index" and "@value" with "%{_index}%", "%loop-value.getKey()%" and "%loop-value.getValue()%" in expression-3
            {_index} = expression-1
            stop loop
        return {_sorted::*}
```

File: src/main/java/org/skriptlang/reflect/syntax/event/EventValuesEntryData.java [hunk 31]
```diff
@@ -0,0 +1,53 @@
...
+import ch.njol.skript.registrations.Classes;
+import org.eclipse.jdt.annotation.Nullable;
+import org.skriptlang.skript.lang.entry.KeyValueEntryData;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class EventValuesEntryData extends KeyValueEntryData<List<ClassInfo<?>>> {
+
+  private static final String listSplitPattern = "\\s*,?\\s+(and|n?or)\\s+|\\s*,\\s*"; // Found in SkriptParser
```
Comment (AyhamAl-Ali) [inline src/main/java/org/skriptlang/reflect/syntax/event/EventValuesEntryData.java:17]: Can we actually make that public on Skript and import it properly here?

File: src/main/java/org/skriptlang/reflect/syntax/event/elements/StructCustomEvent.java [hunk 32]
```diff
@@ -0,0 +1,162 @@
...
+import com.btk5h.skriptmirror.util.SkriptUtil;
+import org.skriptlang.reflect.syntax.CustomSyntaxStructure;
+import org.skriptlang.reflect.syntax.event.BukkitCustomEvent;
+import org.skriptlang.reflect.syntax.event.EventSyntaxInfo;
+import org.skriptlang.reflect.syntax.event.EventTriggerEvent;
+import org.skriptlang.reflect.syntax.event.EventValuesEntryData;
+import org.skriptlang.skript.lang.entry.EntryContainer;
+import org.skriptlang.skript.lang.script.Script;
+
+import java.util.*;
```
Comment (AyhamAl-Ali) [inline src/main/java/org/skriptlang/reflect/syntax/event/elements/StructCustomEvent.java]: Can we avoid that?

File: src/main/resources/plugin.yml [hunk 33]
```diff
@@ -1,10 +1,10 @@
...
-authors: [Bryan Terce <bryan.terce@gmail.com>, TPGamesNL]
-website: https://github.com/TPGamesNL/skript-reflect
+description: Reflection utilities for Skript.
+authors: [Bryan Terce <bryan.terce@gmail.com>, TPGamesNL, 'SkriptLang Team']
+website: https://github.com/SkriptLang/skript-reflect
 api-version: 1.13
 
 main: com.btk5h.skriptmirror.SkriptMirror
-depend: [Skript]
+softdepend: [Skript]
```
Comment (AyhamAl-Ali) [inline src/main/resources/plugin.yml:9]: Shouldn't this stay as depend?

**Review by AyhamAl-Ali — COMMENTED** [review]
Well done 🚀

File: src/main/resources/plugin.yml (hunk 33 above)
Comment (UnderscoreTud) [reply src/main/resources/plugin.yml:9]: We handle it in `onEnable` for a better message as to why the plugin didn't start

File: src/main/java/org/skriptlang/reflect/syntax/event/EventValuesEntryData.java (hunk 31 above)
Comment (UnderscoreTud) [reply src/main/java/org/skriptlang/reflect/syntax/event/EventValuesEntryData.java:17]: For sure, but we can't really do anything about it right now, so it has to stay like this for the time being

Comment (AyhamAl-Ali) [reply src/main/java/org/skriptlang/reflect/syntax/event/EventValuesEntryData.java:17]: Fair enough.

**Review by AyhamAl-Ali — APPROVED** [review]

**Pull request merged**